azurerm/resource_arm_monitor_diagnostic_setting.go

```diff
@@ -0,0 +1,441 @@
+		return fmt.Errorf("Either a `eventhub_authorization_rule_id`, `log_analytics_workspace_id` or `storage_account_id` must be set")
+	}
+
+	// the Azure SDK prefixes the URI with a `/` such this makes a bad request if we don't trim the `/`
+	targetResourceId := strings.TrimPrefix(actualResourceId, "/")
+	_, err := client.CreateOrUpdate(ctx, targetResourceId, properties, name)
+	if err != nil {
+		return fmt.Errorf("Error creating Diagnostics Setting %q for Resource %q: %+v", name, actualResourceId, err)
+	}
+
+	read, err := client.Get(ctx, targetResourceId, name)
+	if err != nil {
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read ID for Monitor Diagnostics %q for Resource ID %q", name, actualResourceId)
+	}
+
+	d.SetId(fmt.Sprintf("%s|%s", actualResourceId, name))
+
+	return resourceArmMonitorDiagnosticSettingRead(d, meta)
+}
+
+func resourceArmMonitorDiagnosticSettingRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorDiagnosticSettingsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseMonitorDiagnosticId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	actualResourceId := id.resourceID
+	targetResourceId := strings.TrimPrefix(actualResourceId, "/")
+	resp, err := client.Get(ctx, targetResourceId, id.name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[WARN] Monitor Diagnostics Setting %q was not found for Resource %q - removing from state!", id.name, actualResourceId)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error retrieving Monitor Diagnostics Setting %q for Resource %q: %+v", id.name, actualResourceId, err)
+	}
+
+	d.Set("name", id.name)
+	d.Set("target_resource_id", id.resourceID)
+
+	d.Set("eventhub_name", resp.EventHubName)
+	d.Set("eventhub_authorization_rule_id", resp.EventHubAuthorizationRuleID)
+	d.Set("log_analytics_workspace_id", resp.WorkspaceID)
+	d.Set("storage_account_id", resp.StorageAccountID)
+
+	if err := d.Set("log", flattenMonitorDiagnosticLogs(resp.Logs)); err != nil {
+		return fmt.Errorf("Error flattening `log`: %+v", err)
+	}
+
+	if err := d.Set("metric", flattenMonitorDiagnosticMetrics(resp.Metrics)); err != nil {
+		return fmt.Errorf("Error flattening `metric`: %+v", err)
+	}
+
+	return nil
+}
+
+func resourceArmMonitorDiagnosticSettingDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorDiagnosticSettingsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseMonitorDiagnosticId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	targetResourceId := id.resourceID
+	actualResourceId := strings.TrimPrefix(targetResourceId, "/")
+	resp, err := client.Delete(ctx, actualResourceId, id.name)
+	if err != nil {
+		if !response.WasNotFound(resp.Response) {
+			return fmt.Errorf("Error deleting Monitor Diagnostics Setting %q for Resource %q: %+v", id.name, targetResourceId, err)
+		}
+	}
+
+	// TODO: we need to poll to ensure this is actually gone
```

Comment:
Is the comment still valid?

Reply:
yep - in some edge-case testing there's some eventual consistency going on here such that we'll want to use a Wait Check Function to ensure it's actually gone imo (sorry thought I'd fixed this prior to pushing)

Reply:
pushed a commit for this